cmd/tealdbg/debugger.go

```diff
@@ -181,25 +221,70 @@ func (s *session) Step() {
 	func() {
 		s.mu.Lock()
 		defer s.mu.Unlock()
-		s.debugConfig = debugConfig{BreakAtLine: stepBreak}
+		s.debugConfig = makeDebugConfig()
+		s.debugConfig.setStepBreak()
 	}()
 
 	s.resume()
 }
 
-func (s *session) Resume() {
-	currentLine := s.line.Load()
+func (s *session) StepOver() {
+	func() {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		// Get the first TEAL opcode in the line
+		currentOp := strings.Fields(s.lines[s.line.Load()])[0]
+		s.debugConfig = makeDebugConfig()
+
+		// Step over a function call (callsub op).
+		if currentOp == "callsub" && s.line.Load() < len(s.breakpoints) {
+			// Set a flag to check if we are in StepOver mode and to
+			// save our initial call depth so we can pass over breakpoints that
+			// are not on the correct call depth.
+			s.debugConfig.setStepOutOver(len(s.callStack))
+			err := s.setBreakpoint(s.line.Load() + 1)
+			if err != nil {
+				s.debugConfig.setStepBreak()
+			}
+		} else {
+			s.debugConfig.setStepBreak()
+		}
+	}()
+	s.resume()
+}
+
+func (s *session) StepOut() {
+	func() {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		s.debugConfig = makeDebugConfig()
+		if len(s.callStack) == 0 {
+			s.debugConfig.setNoBreak()
+		} else {
+			callFrame := s.callStack[len(s.callStack)-1]
+			err := s.setBreakpoint(callFrame.FrameLine + 1)
+			if err != nil {
+				s.debugConfig.setStepBreak()
+			}
+			s.debugConfig.StepOutOver = true
+			s.debugConfig.CallDepth = len(s.callStack)
```

Comment:
@algochoi The _suggestion_ provides a partial answer to the _step out_ issue @jasonpaulos points out in https://github.com/algorand/go-algorand/pull/3653#issuecomment-1083645448.

* Through debugging, it seems clear to me that the PR's `debugConfig.CallDepth` is _not_ compatible with `isBreak`.  During execution, the `callDepth` provided to `isBreak` is `len(s.callStack) - 1`.  Consequently, the intended breakpoint does _not_ trigger.
* The _suggestion_ improves the behavior, but:
  * I think there's at least 1 more problem.
  * I did _not_ run unit tests.  I tested manually as described below.  I recommend running the tests first.  
    * If the tests _don't_ break, then I think the tests are broken or a test is missing.
    * You may also wish to refactor to eliminate duplicate `s.callStack` index management to minimize chance of future divergence.
* I debugged manually using the program @jasonpaulos provided in the linked comment.  Here's the dryrun output I generated by using your tutorial: 
  * [dryrun_txn.msgp.txt](https://github.com/algorand/go-algorand/files/8385599/dryrun_txn.msgp.txt)
  * Github allows limited file extensions for upload.  Remove `.txt` extension.
* Problematic scenario - I discovered it and ran out of time to assess:
  * Open CDT and set breakpoint at line 7 (`dup`).
  * Resume execution.  Confirm CDT jumps to line 7 with call stack size = 2.
  * Step out.  Confirm CDT jumps to line 5 (`return`) with call stack size = 1 (`main`).
  * Resume execution.  Confirm CDT jumps to line 21.
    * I think reaching line 21 is a _problem_.  There's no opcode here.  I think execution _should_ complete.
    * Since I didn't debug, it's unclear to me if it's related to the _suggestion_.  
  * Resume execution.  Confirm CDT completes execution as intended.


```suggestion
			s.debugConfig.CallDepth = len(s.callStack) - 1
```

Reply:
Thanks for the suggestion - I agree it should be `len(s.callStack) - 1`. Without the `-1`, it tries to break too early in the callstack and is most apparent when you try to `StepOut` inside the deepest point of the subroutine. 

Also changed the `pauseOnCompleted` to only be set at the most outer subroutine. This should retain previous functionality before subroutines were introduced.